### PR TITLE
Add --ignore-safety-checks to online upgrade

### LINKED ISSUE #458: Allow for online upgrades when there are ingress nodes
Currently, the upgrade process does not allow for online upgrades when the cluster has an ingress node configured. Ideally, the cluster should allow for an online upgrade when there are more than one nodes and warn the user to ensure that their load balancer is configured properly to route to the active one.

### LINKED ISSUE #619: Problem with online upgrade process
OS: Ubuntu 16.04
etcd: 3 nodes, master: 2 nodes, worker: 3 nodes
Kismatic version: 1.3.0
Version to update: 1.3.3
Install process: 
`kismatic install apply --skip-preflight`

kismatic-cluster.yaml
```
cluster:
  name: kubernetes
  admin_password: ######
  allow_package_installation: true
  disconnected_installation: false
  networking:
    type: routed
    pod_cidr_block: 172.16.0.0/16
    service_cidr_block: 172.17.0.0/16
    policy_enabled: false
    update_hosts_files: false
  certificates:
    expiry: 17520h
  ssh:
    user: kismaticuser
    ssh_key: /opt/kismatic/kismaticuser.key
    ssh_port: 22
docker_registry:
  setup_internal: false
  address: ""
  port: 8443
  CA: ""
etcd:
  expected_count: 3
  nodes:
  - host: "sandbox-etcd-01"
    ip: "10.200.129.118"
    internalip: "10.200.129.118"
  - host: "sandbox-etcd-02"
    ip: "10.200.131.156"
    internalip: "10.200.131.156"
  - host: "sandbox-etcd-03"
    ip: "10.200.130.4"
    internalip: "10.200.130.4"
master:
  expected_count: 2
  load_balanced_fqdn: "10.200.143.251"
  load_balanced_short_name: "10.200.143.251"
  nodes:
  - host: "sandbox-kube-master-01"
    ip: "10.200.128.199"
    internalip: "10.200.128.199"
  - host: "sandbox-kube-master-02"
    ip: "10.200.131.160"
    internalip: "10.200.131.160"
worker:
  expected_count: 3
  nodes:
  - host: "sandbox-kube-worker-01"
    ip: "10.200.133.23"
    internalip: "10.200.133.23"
  - host: "sandbox-kube-worker-02"
    ip: "10.200.128.200"
    internalip: "10.200.128.200"
  - host: "sandbox-kube-worker-03"
    ip: "10.200.133.137"
    internalip: "10.200.133.137"
ingress:
  expected_count: 0
  nodes: []
storage:
  expected_count: 0
  nodes: []
nfs:
  nfs_volume: []
```

Trying to update:
```
Computing upgrade plan==============================================================
Reading plan file                                                               [OK]
Validate SSH connectivity to nodes                                              [OK]

Validate Online Upgrade=============================================================
sandbox-etcd-01 [etcd]                                                          [OK]
sandbox-etcd-02 [etcd]                                                          [OK]
sandbox-etcd-03 [etcd]                                                          [OK]
sandbox-kube-master-01 [master]                                                 [OK]
sandbox-kube-master-02 [master]                                                 [OK]
sandbox-kube-worker-01 [worker]                                                 [ERROR]
- Pod "kube-system/calico-node-07f3l" is using HostPath volume "lib-modules", which is unsafe for upgrades.
- Pod "kube-system/calico-node-07f3l" is using HostPath volume "var-run-calico", which is unsafe for upgrades.
- Pod "kube-system/calico-node-07f3l" is using HostPath volume "cni-bin-dir", which is unsafe for upgrades.
- Pod "kube-system/calico-node-07f3l" is using HostPath volume "cni-net-dir", which is unsafe for upgrades.
- Pod "kube-system/calico-node-07f3l" is using HostPath volume "etcd-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-dns-3895116667-3dxm5" is using HostPath volume "worker-kubeconfig", which is unsafe for upgrades.
- Pod "kube-system/kube-dns-3895116667-3dxm5" is using HostPath volume "ssl-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-01" is using HostPath volume "ssl-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-01" is using HostPath volume "kubeconfig", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-01" is using HostPath volume "etc-kube-ssl", which is unsafe for upgrades.
- Pod "spinnaker/redis-clouddriver-slave-v000-zl6cc" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod "spinnaker/redis-orca-master-v000-dmxvb" is using EmptyDir volume "data", which is unsafe for upgrades.
- The pod "kube-system/kube-proxy-sandbox-kube-worker-01" is not being managed by a controller. Upgrading this node might result in data or availability loss.
- Pod managed by ReplicaSet "spinnaker/redis-front50-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-front50-master-v000" are running on this node.
- Pod managed by ReplicaSet "spinnaker/redis-orca-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-orca-master-v000" are running on this node.
- Pod managed by ReplicationController "spinnaker/spin-deck-v000" is running on this node, and the ReplicationController does not have a replica count greater than 1.
- All the replicas that belong to the ReplicationController "spinnaker/spin-deck-v000" are running on this node.
sandbox-kube-worker-02 [worker]                                                 [ERROR]
- Pod "kube-system/calico-node-7qw76" is using HostPath volume "lib-modules", which is unsafe for upgrades.
- Pod "kube-system/calico-node-7qw76" is using HostPath volume "var-run-calico", which is unsafe for upgrades.
- Pod "kube-system/calico-node-7qw76" is using HostPath volume "cni-bin-dir", which is unsafe for upgrades.
- Pod "kube-system/calico-node-7qw76" is using HostPath volume "cni-net-dir", which is unsafe for upgrades.
- Pod "kube-system/calico-node-7qw76" is using HostPath volume "etcd-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-02" is using HostPath volume "ssl-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-02" is using HostPath volume "kubeconfig", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-02" is using HostPath volume "etc-kube-ssl", which is unsafe for upgrades.
- Pod "kube-system/monitoring-influxdb-3480804314-g8h1t" is using EmptyDir volume "influxdb-storage", which is unsafe for upgrades.
- Pod "spinnaker/redis-clouddriver-slave-v000-b4sr3" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod "spinnaker/redis-fiat-master-v000-4cn66" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod "spinnaker/redis-igor-master-v000-brxkx" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod "spinnaker/redis-rosco-master-v000-lc7l9" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod managed by ReplicaSet "kube-system/heapster-1428305041" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "kube-system/heapster-1428305041" are running on this node.
- The pod "kube-system/kube-proxy-sandbox-kube-worker-02" is not being managed by a controller. Upgrading this node might result in data or availability loss.
- Pod managed by ReplicaSet "kube-system/monitoring-influxdb-3480804314" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "kube-system/monitoring-influxdb-3480804314" are running on this node.
- Pod managed by ReplicaSet "spinnaker/redis-fiat-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-fiat-master-v000" are running on this node.
- Pod managed by ReplicaSet "spinnaker/redis-igor-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-igor-master-v000" are running on this node.
- Pod managed by ReplicaSet "spinnaker/redis-rosco-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-rosco-master-v000" are running on this node.
- Pod managed by ReplicationController "spinnaker/spin-front50-v000" is running on this node, and the ReplicationController does not have a replica count greater than 1.
- All the replicas that belong to the ReplicationController "spinnaker/spin-front50-v000" are running on this node.
sandbox-kube-worker-03 [worker]                                                 [ERROR]
- Pod "kube-system/calico-node-jzr7k" is using HostPath volume "lib-modules", which is unsafe for upgrades.
- Pod "kube-system/calico-node-jzr7k" is using HostPath volume "var-run-calico", which is unsafe for upgrades.
- Pod "kube-system/calico-node-jzr7k" is using HostPath volume "cni-bin-dir", which is unsafe for upgrades.
- Pod "kube-system/calico-node-jzr7k" is using HostPath volume "cni-net-dir", which is unsafe for upgrades.
- Pod "kube-system/calico-node-jzr7k" is using HostPath volume "etcd-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-dns-3895116667-m9sbc" is using HostPath volume "worker-kubeconfig", which is unsafe for upgrades.
- Pod "kube-system/kube-dns-3895116667-m9sbc" is using HostPath volume "ssl-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-03" is using HostPath volume "ssl-certs", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-03" is using HostPath volume "kubeconfig", which is unsafe for upgrades.
- Pod "kube-system/kube-proxy-sandbox-kube-worker-03" is using HostPath volume "etc-kube-ssl", which is unsafe for upgrades.
- Pod "spinnaker/redis-clouddriver-master-v000-n727v" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod "spinnaker/redis-clouddriver-slave-v000-d2sr7" is using EmptyDir volume "data", which is unsafe for upgrades.
- Pod "spinnaker/redis-gate-master-v000-xfjsm" is using EmptyDir volume "data", which is unsafe for upgrades.
- The pod "kube-system/kube-proxy-sandbox-kube-worker-03" is not being managed by a controller. Upgrading this node might result in data or availability loss.
- Pod managed by ReplicaSet "spinnaker/redis-clouddriver-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-clouddriver-master-v000" are running on this node.
- Pod managed by ReplicaSet "spinnaker/redis-gate-master-v000" is running on this node, and the ReplicaSet does not have a replica count greater than 1.
- All the replicas that belong to the ReplicaSet "spinnaker/redis-gate-master-v000" are running on this node.
- Pod managed by ReplicationController "spinnaker/spin-echo-event-v000" is running on this node, and the ReplicationController does not have a replica count greater than 1.
- All the replicas that belong to the ReplicationController "spinnaker/spin-echo-event-v000" are running on this node.
- Pod managed by ReplicationController "spinnaker/spin-echo-sched-v000" is running on this node, and the ReplicationController does not have a replica count greater than 1.
- All the replicas that belong to the ReplicationController "spinnaker/spin-echo-sched-v000" are running on this node.
- Pod managed by ReplicationController "spinnaker/spin-igor-v000" is running on this node, and the ReplicationController does not have a replica count greater than 1.
- All the replicas that belong to the ReplicationController "spinnaker/spin-igor-v000" are running on this node.
Unable to perform an online upgrade due to the unsafe conditions detected.
```
The question is how to install Kubernetes via Kismatic, for safe upgrade it?

### LINKED ISSUE #636: Fix Online Upgrade Process
As a first step, we are going to add a flag when doing `upgrade online` that allows to `--skip-safety-checks`.

This should address #619 and #458

### PR DESCRIPTION
**Pull request by dkoshkin**
Fixes #636 , Fixes #619, Fixes #458 

* Added `--skip-safety-checks` flag to online upgrades. The safety-checks will still run however the upgrade will continue even if the tests fail.
* Always drain nodes, even on offline upgrade.
* Added integration tests for online upgrades.

### CODE REVIEW

**Comment by ikester** [comment]
As discussed briefly offline, I like the idea of running the checks regardless, to record the issues that are being ignored/bypassed.

Consider renaming the flag to `--ignore-safety-checks` or `--bypass-safety-checks`

**Comment by dkoshkin** [comment]
I like `--ignore-safety-checks`, going with that.

File: integration/upgrade_test.go [hunk 1]
```diff
@@ -234,9 +420,28 @@ func extractCurrentKismaticInstaller() {
...
 	FailIfError(err)
 }
-func upgradeCluster() {
+func upgradeCluster(online bool) {
 	// Perform upgrade
 	cmd := exec.Command("./kismatic", "upgrade", "offline", "-f", "kismatic-testing.yaml")
+	if online {
+		// wait for kuberang pods to shut down
+		time.Sleep(60 * time.Second)
+		// if err := retry.WithBackoff(func() error {
```
Comment (alexbrand) [inline integration/upgrade_test.go]: Do we need this code commented out?

File: integration/upgrade_test.go [hunk 2]
```diff
@@ -234,9 +420,28 @@ func extractCurrentKismaticInstaller() {
 	err = extractCurrentKismatic(pwd)
 	FailIfError(err)
 }
-func upgradeCluster() {
+func upgradeCluster(online bool) {
 	// Perform upgrade
 	cmd := exec.Command("./kismatic", "upgrade", "offline", "-f", "kismatic-testing.yaml")
+	if online {
+		// wait for kuberang pods to shut down
+		time.Sleep(60 * time.Second)
```
Comment (alexbrand) [inline integration/upgrade_test.go]: Let's go with this for now, but I wonder if we'll need a better signal.

File: integration/upgrade_test.go (hunk 1 above)
Comment (dkoshkin) [inline integration/upgrade_test.go]: Yeah I spent a few hours trying to get this to work, but it didn't, I might want to come back to this at some point in the future.

Comment (alexbrand) [inline integration/upgrade_test.go]: Got it!

File: pkg/cli/upgrade.go [hunk 3]
```diff
@@ -270,20 +273,44 @@ func upgradeNodes(out io.Writer, plan install.Plan, opts upgradeOpts, nodesNeedU
...
-						return errors.New("Unable to perform an online upgrade due to the unsafe conditions detected.")
+
+		// if --skipSafetyChecks still want to run and print the checks, just ignore them
+		if opts.skipSafetyChecks {
+			if len(unsafeNodes) > 0 {
+				util.PrettyPrintWarn(out, "\nIgnoring safety checks and continuing with the upgrade")
+			}
+		} else {
+			var safetyErr error
+			// If we found any unsafe nodes, and we are not doing a partial upgrade, exit.
```
Comment (alexbrand) [inline pkg/cli/upgrade.go]: We can probably update/remove this comment, as we are not doing this anymore.

**Pull request merged**